src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs

```diff
@@ -245,32 +244,37 @@ public override TaintedDataAbstractValue VisitObjectCreation(IObjectCreationOper
                     }
                     else if (this.DataFlowAnalysisContext.SourceInfos.IsSourceMethod(
                         method,
-                        out evaluateWithPointsToAnalysis,
-                        out evaluateWithValueContentAnalysis))
+                        visitedArguments,
+                        (originalOperation as IInvocationOperation).Arguments.Select(o => GetPointsToAbstractValue(o.Value)),
+                        (originalOperation as IInvocationOperation).Arguments.Select(o => GetValueContentAbstractValue(o.Value)),
+                        out taintedTargets))
                     {
-                        ImmutableArray<IArgumentOperation> argumentOperation = (originalOperation as IInvocationOperation).Arguments;
-                        IEnumerable<PointsToAbstractValue> pointsToAnalysisResult = argumentOperation.Select(o => GetPointsToAbstractValue(o.Value));
-                        IEnumerable<ValueContentAbstractValue> valueContentAnalysisResult = argumentOperation.Select(o => GetValueContentAbstractValue(o.Value));
-                        if ((evaluateWithPointsToAnalysis != null && evaluateWithPointsToAnalysis.Any(o => o(pointsToAnalysisResult)))
-                            || (evaluateWithValueContentAnalysis != null && evaluateWithValueContentAnalysis.Any(o => o(pointsToAnalysisResult, valueContentAnalysisResult))))
+                        foreach (var taintedTarget in taintedTargets)
                         {
-                            return TaintedDataAbstractValue.CreateTainted(method, originalOperation.Syntax, this.OwningSymbol);
+                            if (taintedTarget != "RETURN")
+                            {
+                                IArgumentOperation argumentOperation = visitedArguments.FirstOrDefault(o => o.Parameter.Name == taintedTarget);
+                                if (argumentOperation != null && AnalysisEntityFactory.TryCreate(argumentOperation, out AnalysisEntity analysisEntity))
+                                {
+                                    IParameterReferenceOperation argumentValue = argumentOperation.Value as IParameterReferenceOperation;
+                                    this.CurrentAnalysisData.SetAbstractValue(analysisEntity, TaintedDataAbstractValue.CreateTainted(argumentValue.Parameter, argumentValue.Syntax, this.OwningSymbol));
+                                }
+                            }
                         }
 
-                        return TaintedDataAbstractValue.NotTainted;
+                        if (taintedTargets.Contains("RETURN"))
+                        {
+                            return TaintedDataAbstractValue.CreateTainted(method, originalOperation.Syntax, this.OwningSymbol);
+                        }
+                        else
```

Comment:
Since the foreach already enumerates taintedTargets, it should have a flag that indicates the return taintedTarget was found.  Then the if could check the flag instead of enumerating again.